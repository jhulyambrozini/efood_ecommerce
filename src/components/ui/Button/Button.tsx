import { ButtonContainer } from './styles'

export type ButtonProps = {
  type: 'button' | 'submit'
  background: 'primary' | 'secundary'
  text?: string
  title: string
  onClick?: React.MouseEventHandler<HTMLButtonElement>
}

const Button = ({ type, background, text, title, onClick }: ButtonProps) => (
  <ButtonContainer
    background={background}
    title={title}
    type={type}
    onClick={onClick}
  >
    {text ? text : title}
  </ButtonContainer>
)

export default Button
