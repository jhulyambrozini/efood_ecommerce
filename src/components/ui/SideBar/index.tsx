import { useDispatch, useSelector } from 'react-redux'

import { Overlay } from '../../../styles'
import { Container, AsideContainer, ButtonClose } from './styles'
import closeIcon from '../../../assets/images/close.png'

import { changeComponent, closeSideBar } from '../../../store/reducers/sideBar'
import { RootReducer } from '../../../store'
import Cart from '../../view/Cart'
import Form from '../../view/Form/Form'

const SideBar = () => {
  const dispatch = useDispatch()
  const { component } = useSelector((state: RootReducer) => state.sideBar)
  const { itemsCart } = useSelector((state: RootReducer) => state.cart)

  const switchComponents = () => {
    switch (component) {
      case 'cart':
        return <Cart itemsCart={itemsCart} />
      case 'form':
        return <Form />
      default:
        return null
    }
  }

  const toClose = () => {
    if (component === 'form') {
      dispatch(closeSideBar())
      dispatch(changeComponent('cart'))
    } else {
      dispatch(closeSideBar())
    }
  }
  return (
    <Container>
      <Overlay onClick={toClose} />
      <AsideContainer>
        <ButtonClose onClick={toClose} title="Fechar carrinho">
          <img src={closeIcon} alt="Icone de fechar" />
        </ButtonClose>

        {switchComponents()}
      </AsideContainer>
    </Container>
  )
}

export default SideBar
