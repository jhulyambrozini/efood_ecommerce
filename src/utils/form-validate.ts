import { FormikProps } from 'formik'

export const getMessageError = (
  field: string,
  form: FormikProps<FormValues>,
  message?: string
) => {
  const isTouched = field in form.touched
  const isInvalid = field in form.errors

  if (isTouched && isInvalid && message) return message
  return ''
}
