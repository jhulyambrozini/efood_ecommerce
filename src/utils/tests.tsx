import { PreloadedState } from '@reduxjs/toolkit'
import { RenderOptions, render } from '@testing-library/react'
import { PropsWithChildren, ReactElement } from 'react'
import { RootState, AppStore, configStore } from '../store'
import { Provider } from 'react-redux'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProvider(
  element: ReactElement,
  {
    preloadedState = {},
    store = configStore(preloadedState),
    ...moreOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(element, {
      wrapper: Wrapper,
      ...moreOptions
    })
  }
}
