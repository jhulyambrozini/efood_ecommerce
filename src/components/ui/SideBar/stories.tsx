import type { Meta } from '@storybook/react'

import SideBar from '.'

const meta = {
  title: 'ui/SideBar',
  component: SideBar,
  parameters: {
    layout: 'fullscreen'
  }
} satisfies Meta<typeof SideBar>

// eslint-disable-next-line react-refresh/only-export-components
export default meta

export const Template = () => {
  return <SideBar />
}
