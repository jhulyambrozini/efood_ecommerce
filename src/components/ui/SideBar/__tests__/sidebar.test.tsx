import { fireEvent, screen } from '@testing-library/dom'
import SideBar from '..'
import { renderWithProvider } from '../../../../utils/tests'
import Button from '../../Button'

describe('<Sidebar />', () => {
  it('should render correctly', () => {
    renderWithProvider(<SideBar />)
    expect(screen.getByText(/O carrinho estÃ¡ vazio/i)).toBeInTheDocument()
  })

  it('should call a function when pressing the close button', () => {
    const handleClose = jest.fn()
    renderWithProvider(<SideBar />)
    const button = screen.getByRole('button', { name: 'Icone de fechar' })
    expect(button).toBeInTheDocument()

    button.onclick = handleClose

    fireEvent.click(button)

    expect(handleClose).toHaveBeenCalled()
  })

  it('should call a function when pressing the overlay container', () => {
    const handleCloseOverlay = jest.fn()
    renderWithProvider(<SideBar />)
    const overlayContainer = screen.getByRole('overlay-container')
    expect(overlayContainer).toBeInTheDocument()

    overlayContainer.onclick = handleCloseOverlay

    fireEvent.click(overlayContainer)

    expect(handleCloseOverlay).toHaveBeenCalled()
  })

  it('should render with a Cart component', () => {
    renderWithProvider(<SideBar />, {
      preloadedState: {
        sideBar: {
          component: 'cart',
          sideBarIsOpen: true
        },
        cart: {
          itemsCart: [
            {
              foto: 'img/img.png',
              preco: 12.7,
              id: 1,
              nome: 'comida 1',
              descricao: 'descricao 1',
              porcao: '2'
            }
          ]
        }
      }
    })

    const buttonContinueWithDelivery = screen.getByRole('button', {
      name: /continuar com a entrega/i
    })
    expect(buttonContinueWithDelivery).toBeInTheDocument()
  })

  it('should render with a Form component', () => {
    renderWithProvider(<SideBar />, {
      preloadedState: {
        sideBar: {
          component: 'form',
          sideBarIsOpen: true
        }
      }
    })

    const headingDelivery = screen.getByRole('heading', { name: /entrega/i })
    expect(headingDelivery).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = renderWithProvider(<SideBar />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
