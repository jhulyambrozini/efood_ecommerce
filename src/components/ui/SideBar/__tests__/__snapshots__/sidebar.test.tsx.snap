// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should match snapshot 1`] = `
<div
  class="sc-kpDqfm DGtKj"
>
  <div
    class="sc-gEvEer cSvWod"
    role="overlay-container"
  />
  <aside
    class="sc-dAlyuH cGdzvi"
  >
    <button
      class="sc-eqUAAy sc-jlZhew KzdVX dUHYKG"
      title="Fechar carrinho"
    >
      <img
        alt="Icone de fechar"
        src="test-file-stub"
      />
    </button>
    <div>
      <p
        class="empty-text"
        style="color: rgba(255, 255, 255, 1);"
      >
        O carrinho est√° vazio, dicione ao menos um item para continuar com a compra.
      </p>
    </div>
  </aside>
</div>
`;
